query_optimizer/ExecutionGenerator.cpp

```diff
@@ -740,6 +805,202 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
+        const attribute_id probe_partition_attr = probe_partition_scheme_header.getPartitionAttributeId();
+
+        probe_needs_repartition =
+            (find(probe_attribute_ids.begin(), probe_attribute_ids.end(), probe_partition_attr) ==
+                probe_attribute_ids.end());
+        break;
+      }
+    }
+  } else if (build_partition_scheme) {
+    const PartitionSchemeHeader &build_partition_scheme_header = build_partition_scheme->getPartitionSchemeHeader();
+    switch (build_partition_scheme_header.getPartitionType()) {
+      case PartitionSchemeHeader::PartitionType::kRange:
+        build_needs_repartition = true;
+        break;
+      case PartitionSchemeHeader::PartitionType::kHash: {
+        const attribute_id build_partition_attr = build_partition_scheme_header.getPartitionAttributeId();
+        build_needs_repartition =
+            (find(build_attribute_ids.begin(), build_attribute_ids.end(), build_partition_attr) ==
+                build_attribute_ids.end());
+        break;
+      }
+    }
+  }
+
+  if (needs_swap) {
+    swap(probe_physical, build_physical);
+    swap(probe_attribute_ids, build_attribute_ids);
+    swap(any_probe_attributes_nullable, any_build_attributes_nullable);
+    swap(left_join_attributes, right_join_attributes);
+    swap(probe_operator_info, build_relation_info);
+    swap(probe_relation, build_relation);
+    swap(probe_partition_scheme, build_partition_scheme);
+    swap(probe_needs_repartition, build_needs_repartition);
+  }
+
+  unique_ptr<PartitionScheme> probe_repartition_scheme;
+  if (build_needs_repartition) {
+    const size_t repartition_attr_index =
+        chooseBestRepartitionAttributeIndex(build_relation->getStatistics(), build_attribute_ids);
+    auto build_repartition_scheme_header =
+        make_unique<HashPartitionSchemeHeader>(FLAGS_num_repartitions,
+                                               build_attribute_ids[repartition_attr_index]);
+    auto build_repartition_scheme = make_unique<PartitionScheme>(build_repartition_scheme_header.release());
+
+    build_partition_scheme = build_repartition_scheme.get();
+
+    if (build_relation_info->isStoredRelation()) {
+      THROW_SQL_ERROR() << "Re-partition for the base build table is not supported";
+    } else {
+      S::InsertDestination *build_insert_destination_proto =
+          query_context_proto_->mutable_insert_destinations(build_relation_info->output_destination_index);
+
+      build_insert_destination_proto->set_insert_destination_type(S::InsertDestinationType::PARTITION_AWARE);
+      build_insert_destination_proto->MutableExtension(S::PartitionAwareInsertDestination::partition_scheme)
+        ->MergeFrom(build_repartition_scheme->getProto());
+
+      CatalogRelation *mutable_build_relation =
+          catalog_database_->getRelationByIdMutable(build_relation->getID());
+      mutable_build_relation->setPartitionScheme(build_repartition_scheme.release());
```

Comment:
We modify the existing `SelectOperator` to have a new `PartitionAwareInsertDestination` for repartition, if the input relation is not a base table. Otherwise, we will create one, but not part of this PR.